packages/java-parser/src/tokens.js

```diff
@@ -32,17 +33,46 @@ FRAGMENT("HexDigits", "{{HexDigit}}(({{HexDigit}}|'_')*{{HexDigit}})?");
 FRAGMENT("FloatTypeSuffix", "[fFdD]");
 FRAGMENT("LineTerminator", "(\\x0A|(\\x0D(\\x0A)?))");
 
+function matchJavaIdentifier(char, startOffset) {
+  let endOffset = startOffset;
+  let charCode = char.codePointAt(endOffset);
+
+  // We verifiy if the first character is from one of these categories
+  // Corresponds to the isJavaIdentifierStart function from Java
+  if (chars.firstIdentChar.has(charCode)) {
+    endOffset++;
+    charCode = char.charCodeAt(endOffset);
+  }
+
+  // We verify if the remaining characters is from one of these categories
+  // Corresponds to the isJavaIdentifierPart function from Java
+  while (chars.restIdentChar.has(charCode)) {
+    endOffset++;
+    charCode = char.codePointAt(endOffset);
+  }
+
+  // No match, must return null to conform with the RegExp.prototype.exec signature
+  if (endOffset === startOffset) {
+    return null;
+  }
+  const matchedString = char.substring(startOffset, endOffset);
```

Comment:
I do not think the offsets are correct for unicode characters outside UTF16
Because:
```
"💩".length // === 2
```

Reply:
If any of the Identifier characters are encoded using surrogate pairs we will need to take this into account.